R/module_teal_lockfile.R

```diff
@@ -0,0 +1,195 @@
+#' Generate lockfile for application's environment reproducibility
+#'
+#' @param lockfile_path (`character`) path to the lockfile (`"teal_app.lock"`).
+#'
+#' @section Different ways of creating lockfile:
+#' `teal` leverages [renv::snapshot()], which offers multiple methods for lockfile creation.
+#'
+#' - User-specified:
+#'     - **Pre-computed lockfile**: Users can provide their own pre-computed lockfile by specifying the path via
+#'     `teal.renv.lockfile` option. Automatic lockfile computation is skipped in such case.
+#' - Automatically computed:
+#'     - **Working directory lockfile**: If `teal.renv.lockfile` is not set, `teal` will, by default, create an
+#'      `implicit` type lockfile that uses `renv::dependencies()` to detect all R packages in the current project's
+#'      working directory.
+#'     - **`DESCRIPTION`-based lockfile**: To generate a lockfile based on a `DESCRIPTION` file in your working
+#'     directory, set `renv::settings$snapshot.type("explicit")`. The naming convention for `type` follows
+#'     `renv::snapshot()`. For the `"explicit"` type, refer to `renv::settings$package.dependency.fields()` for the
+#'     `DESCRIPTION` fields included in the lockfile.
+#'     - **Custom files-based lockfile**: To specify custom files as the basis for the lockfile, set
+#'     `renv::settings$snapshot.type("custom")` and configure the `renv.snapshot.filter` option.
+#'
+#' @section lockfile usage:
+#' After creating the lockfile, you can restore the application environment using `renv::restore()`.
+#'
+#' @seealso [renv::snapshot()], [renv::restore()].
+#'
+#' @return `NULL`
+#'
+#' @name module_teal_lockfile
+#' @rdname module_teal_lockfile
+#'
+#' @keywords internal
+NULL
+
+#' @rdname module_teal_lockfile
+ui_teal_lockfile <- function(id) {
+  if (!isTRUE(getOption("teal.renv.enable"))) {
+    return(NULL)
+  } else if (!.is_lockfile_deps_installed()) {
+    warning("Lockfile feature disabled. `mirai` and `renv` packages must be installed.")
+    return(NULL)
+  }
```

Comment:
I think it would be better to keep ui independent (i.e. always return disabled link) and handle soft dependencies in server only.
- enable option at the beggining where it is currently
- soft deps in `.teal_lockfile_process_invoke` - i.e. return empty `process` object if any package is missing

Reply:
> I think it would be better to keep ui independent (i.e. always return disabled link)

I don't like this proposition. I don't see the reason why app-user would like to see disabled button when packages are missing. What if `teal.lockfile.enable = FALSE`, should disabled download button displayed also? 

I don't see the point to run `server` and make some extra `if`s there just to disable download button. Creating empty `process` will require specific code to handle this "exception". I consider this too complicated (more complicated that simple ifs in both UI and server). This is how I view this now.

<img width="542" alt="image" src="https://github.com/user-attachments/assets/58323983-3653-4ec0-a9ae-e77cd63c2997">

Reply:
I was more concerned about messy implementation of multiple ifs here and there but if you put this in that way (I mean the decision tree from the above) then it's going to be good. I think it's connected with https://github.com/insightsengineering/teal/pull/1263/files#r1745602447 so once implemented let me have another look at this.